src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp

```diff
@@ -448,10 +706,95 @@ static inline std::tuple<bool, // is valid
+                continue;
+            if(cfg.precision != precision)
+                continue;
+            // nxe==0 case do not have padding function
+            if(cfg.nxe == 0)
+                continue;
+
+            const auto& gemm_m_per_block = cfg.gemm_m_per_block;
+            const auto& gemm_n_per_block = cfg.gemm_n_per_block;
+            const auto& gemm_k_per_block = cfg.gemm_k_per_block;
+
+            const auto b      = (ho * wo + cfg.nxb - 1) / cfg.nxb * cfg.nxb;
+            const auto gemm_k = n * b;
+            if(gemm_k % gemm_k_per_block != 0)
+            {
+                continue;
+            }
+            // do not need to check gemm_n when n_c0 == 1; n_c0 = t_c0 * c_c0
+            if(cfg.tensor_b_thread_lengths[2] * cfg.tensor_b_cluster_lengths[2] > 1)
+            {
+
+                if(c % gemm_n_per_block != 0)
+                {
+                    continue;
+                }
+            }
+
+            if(cfg.tensor_a_thread_lengths[2] * cfg.tensor_a_thread_lengths[3] > 1)
+            {
+                if(gemm_m % gemm_m_per_block != 0)
+                {
+                    continue;
+                }
+            }
+
+            if(wo % cfg.tensor_b_thread_lengths[1] != 0)
+            {
+                continue;
+            }
+            if(cfg.tensor_b_thread_lengths[1] != 1 && (pad_h != 0 || pad_w != 0))
+            {
+                continue;
+            }
+
+            int gemm_k_global_split = if_gemm_k_global_split(
+                ctx, gemm_m_per_block, gemm_n_per_block, gemm_k_per_block, b);
+            int log2_gemm_k_global_splits = 0;
+            int grid_size                 = integer_divide_ceil(gemm_m, gemm_m_per_block) *
+                            integer_divide_ceil(gemm_n, gemm_n_per_block);
+            int block_size = cfg.GetBlockSize();
+            log2_gemm_k_global_splits =
+                ComputeLog2GemmKGlobalSplits(grid_size, max_grid_size, n, b, gemm_k_per_block);
+            if(gemm_k_global_split == 0)
+                log2_gemm_k_global_splits = 0;
+
+            grid_size = grid_size << log2_gemm_k_global_splits;
+
+            is_valid                      = true;
+            sel_block_size                = block_size;
+            sel_grid_size                 = grid_size;
+            sel_index                     = cfg_index;
+            sel_log2_gemm_k_global_splits = log2_gemm_k_global_splits;
+            break;
+        }
+    }
+
     return std::make_tuple(
         is_valid, sel_index, sel_block_size, sel_grid_size, sel_log2_gemm_k_global_splits);
 }
 
+size_t ConvAsmImplicitGemmGTCDynamicWrwXdlops::GetWorkspaceSize(const ConvolutionContext& ctx) const
+{
+    int k       = ctx.n_inputs;
+    int c       = ctx.n_outputs;
+    int y       = ctx.kernel_size_h;
+    int x       = ctx.kernel_size_w;
+    int ngroups = ctx.group_counts;
+
+    return ngroups * (k / ngroups) * (c / ngroups) * y * x * sizeof(float);
```

Comment:
Why GetSolution returns different workspace size?

Reply:
@atamazov Do you mean the reason why fp32 datatype has different workspace size compared to fp16?
As we use atomic add to split `n*ho*wo`, but gfx908 do not have fp16 atomic add instruction. So for fp32, we directly do atomic add. And for fp16 datatype, we need to create a buffer to do atomic add in fp32, and then cast the buffer to fp16 datatype.

Reply:
No. AFAICS under certain conditions GetWorkspaceSize and GetSolution may return different workspace sizes. Am I right?

Reply:
`if(ctx.IsFp32())
       result.workspce_sz = 0;
 else if(ctx.IsFp16())
       result.workspce_sz = log2_gemm_k_global_splits > 0 ? GetWorkspaceSize(ctx) : 0;`

Yes, I do workspace calculation by this way in GetSolution.
So, is it better to move these lines into GetWorkspaceSize to keep GetWorkspaceSize and GetSolution have the same return value?

Reply:
Changed, the 2 function return the same value.

Reply:
Yes, thanks.

GetWorkspaceSize and GetSolution must return the same workspace size for all applicable problem configs. This is one of the basic guarantees that all solvers must provide.